File: core/src/main/java/com/ariasaproject/advancerofrpg/physics/box2d/Fixture.java
package com.ariasaproject.advancerofrpg.physics.box2d;

import com.ariasaproject.advancerofrpg.math.Vector2;
import com.ariasaproject.advancerofrpg.physics.box2d.Shape.Type;

public class Fixture {
    static {
        initialize();
    }

    private final Filter filter = new Filter();
    private final long[] shapeTmp = new long[2];
    private final short[] tmp = new short[3];
    protected long addr; // used in native
    protected Shape shape;
    protected Object userData;
    private Body body;
    private boolean dirtyFilter = true;

    protected Fixture(Body body, long addr) {
        this.body = body;
        this.addr = addr;
    }

    static native void initialize();

    protected void reset(Body body, long addr) {
        this.body = body;
        this.addr = addr;
        this.shape = null;
        this.userData = null;
        this.dirtyFilter = true;
    }

    public Type getType() {
        int type = jniGetType();
        switch (type) {
            case 0:
                return Type.Circle;
            case 1:
                return Type.Edge;
            case 2:
                return Type.Polygon;
            case 3:
                return Type.Chain;
            default:
                throw new RuntimeException("Unknown shape type!");
        }
    }

    private native int jniGetType();

    public Shape getShape() {
        if (shape == null) {
            jniGetShape(shapeTmp);
            if (shapeTmp[0] == 0)
                throw new RuntimeException("Null shape address!");
            switch ((int) shapeTmp[1]) {
                case 0:
                    shape = new CircleShape(shapeTmp[0]);
                    break;
                case 1:
                    shape = new EdgeShape(shapeTmp[0]);
                    break;
                case 2:
                    shape = new PolygonShape(shapeTmp[0]);
                    break;
                case 3:
                    shape = new ChainShape(shapeTmp[0]);
                    break;
                default:
                    throw new RuntimeException("Unknown shape type!");
            }
        }

        return shape;
    }

    private native long jniGetShape(long[] tmp);

    public native boolean isSensor();

    public native void setSensor(boolean sensor);

    private native void jniSetFilterData(short categoryBits, short maskBits, short groupIndex);

    public Filter getFilterData() {
        if (dirtyFilter) {
            jniGetFilterData(tmp);
            filter.maskBits = tmp[0];
            filter.categoryBits = tmp[1];
            filter.groupIndex = tmp[2];
            dirtyFilter = false;
        }
        return filter;
    }

    public void setFilterData(Filter filter) {
        jniSetFilterData(filter.categoryBits, filter.maskBits, filter.groupIndex);
        this.filter.set(filter);
        dirtyFilter = false;
    }

    private native void jniGetFilterData(short[] filter);

    public native void refilter();

    public Body getBody() {
        return body;
    }

    public boolean testPoint(Vector2 p) {
        return testPoint(p.x, p.y);
    }

    public native boolean testPoint(float x, float y);

// const b2Body* GetBody() const;
//
// /// Get the next fixture in the parent body's fixture list.
// /// @return the next shape.
// b2Fixture* GetNext();
// const b2Fixture* GetNext() const;
//
// /// Get the user data that was assigned in the fixture definition. Use this to
// /// store your application specific data.
// void* GetUserData() const;
//
// /// Set the user data. Use this to store your application specific data.
// void SetUserData(void* data);
//
// /// Cast a ray against this shape.
// /// @param output the ray-cast results.
// /// @param input the ray-cast input parameters.
// bool RayCast(b2RayCastOutput* output, const b2RayCastInput& input) const;
//
// /// Get the mass data for this fixture. The mass data is based on the density and
// /// the shape. The rotational inertia is about the shape's origin. This operation
// /// may be expensive.
// void GetMassData(b2MassData* massData) const;

    public native float getDensity();

    public native void setDensity(float density);

    public native float getFriction();

    public native void setFriction(float friction);

    public native float getRestitution();

    public native void setRestitution(float restitution);
// /// Get the fixture's AABB. This AABB may be enlarge and/or stale.
// /// If you need a more accurate AABB, compute it using the shape and
// /// the body transform.
// const b2AABB& GetAABB() const;

    public Object getUserData() {
        return userData;
    }

    public void setUserData(Object userData) {
        this.userData = userData;
    }
}

File: native-core/box2d/Fixture.cpp
#include "Fixture.h"
#include "Dynamics/b2Fixture.h"

//java value id
static jfieldID fixturePtr;

Fixture_M(void, initialize)(JNIEnv *env, jclass clazz) {
    //initialize field ID of JNI
    fixturePtr = env->GetFieldID(clazz, "addr", "J");
}

Fixture_M(jint, jniGetType)(JNIEnv *env, jobject obj) {
    b2Shape::Type type = ((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetType();
    switch (type) {
        case b2Shape::e_circle:
            return 0;
        case b2Shape::e_edge:
            return 1;
        case b2Shape::e_polygon:
            return 2;
        case b2Shape::e_chain:
            return 3;
        default:
            return -1;
    }
}

Fixture_M(void, jniGetShape)(JNIEnv *env, jobject obj, jlongArray vals) {
    jlong *data = (jlong *) env->GetPrimitiveArrayCritical(vals, 0);
    data[0] = (jlong) ((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetShape();
    switch (((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetShape()->GetType()) {
        case b2Shape::e_circle:
            data[1] = 0;
            break;
        case b2Shape::e_edge:
            data[1] = 1;
            break;
        case b2Shape::e_polygon:
            data[1] = 2;
            break;
        case b2Shape::e_chain:
            data[1] = 3;
            break;
        default:
            data[0] = -1;
            break;
    }
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

Fixture_M(void, setSensor)(JNIEnv *env, jobject obj, jboolean sensor) {
    ((b2Fixture *) env->GetLongField(obj, fixturePtr))->SetSensor(sensor);
}

Fixture_M(jboolean, isSensor)(JNIEnv *env, jobject obj) {
    return ((b2Fixture *) env->GetLongField(obj, fixturePtr))->IsSensor();
}

Fixture_M(void, jniSetFilterData)(JNIEnv *env, jobject obj, jshort categoryBits, jshort maskBits,
                                  jshort groupIndex) {
    b2Filter filter;
    filter.categoryBits = categoryBits;
    filter.maskBits = maskBits;
    filter.groupIndex = groupIndex;
    ((b2Fixture *) env->GetLongField(obj, fixturePtr))->SetFilterData(filter);
}

Fixture_M(void, jniGetFilterData)(JNIEnv *env, jobject obj, jshortArray vals) {
    unsigned short *data = (unsigned short *) env->GetPrimitiveArrayCritical(vals, 0);
    b2Filter f = ((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetFilterData();
    data[0] = f.maskBits;
    data[1] = f.categoryBits;
    data[2] = f.groupIndex;
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

Fixture_M(void, refilter)(JNIEnv *env, jobject obj) {
    ((b2Fixture *) env->GetLongField(obj, fixturePtr))->Refilter();
}

Fixture_M(jboolean, testPoint)(JNIEnv *env, jobject obj, jfloat x, jfloat y) {
    return ((b2Fixture *) env->GetLongField(obj, fixturePtr))->TestPoint(b2Vec2(x, y));
}

Fixture_M(void, setDensity)(JNIEnv *env, jobject obj, jfloat density) {
    ((b2Fixture *) env->GetLongField(obj, fixturePtr))->SetDensity(density);
}

Fixture_M(jfloat, getDensity)(JNIEnv *env, jobject obj) {
    return ((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetDensity();
}

Fixture_M(jfloat, getFriction)(JNIEnv *env, jobject obj) {
    return ((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetFriction();
}

Fixture_M(void, setFriction)(JNIEnv *env, jobject obj, jfloat friction) {
    ((b2Fixture *) env->GetLongField(obj, fixturePtr))->SetFriction(friction);
}

Fixture_M(jfloat, getRestitution)(JNIEnv *env, jobject obj) {
    return ((b2Fixture *) env->GetLongField(obj, fixturePtr))->GetRestitution();
}

Fixture_M(void, setRestitution)(JNIEnv *env, jobject obj, jfloat restitution) {
    ((b2Fixture *) env->GetLongField(obj, fixturePtr))->SetRestitution(restitution);

}

